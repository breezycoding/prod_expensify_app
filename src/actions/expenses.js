import uuid from "uuid";
//ADD EXPENSE
export const add_expense = ({ description = "", note = "", amount = 0, created_at = 0} = {}) => ({
	type: "ADD_EXPENSE",
	expense:{
		id:uuid(),
		description,
		note,
		amount, 
		created_at
	}
});

// REMOVE EXPENSE
export const remove_expense = ({ id } = {}) => ({
	type: "REMOVE_EXPENSE",
	id
});

//EDIT EXPENSE
export const edit_expense = (id, updates) => ({
	type: "EDIT_EXPENSE",
	id,
	updates
});