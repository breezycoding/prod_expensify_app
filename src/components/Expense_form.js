import React from "react";
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates";

export default class Expense_form extends React.Component{
	constructor(props){
		super(props);
		this.state = {
			description: props.expense ? props.expense.description : "",
			note: props.expense ? props.expense.note : "",
			amount: props.expense ? (props.expense.amount / 100).toString() : "",
			created_at: props.expense ? moment(props.expense.created_at) : moment(),
			calendar_focused: false,
			error_message: ""
		}
	}
	
	on_description_change = (event) => {
		const description = event.target.value;
		this.setState(()=> ({description}));
	}
	
	on_note_change = (event) => {
		const note = event.target.value;
		this.setState(()=> ({note}));
	}
	
	on_amount_change = (event) => {
		const amount = event.target.value;
		if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
			this.setState(() => ({amount}));
		}
	}
	
	on_date_change = (created_at) => {
		if(created_at){
			this.setState(() => ({created_at}));
		}
	}
	
	on_focus_change = ({focused}) => {
		this.setState(() => ({calendar_focused:focused}));
	}
	
	on_submit = (event) => {
		event.preventDefault();
		if(!this.state.description || !this.state.amount){
			this.setState(()=>({error_message:"Please provide description and amount"}));
		}else{
			this.setState(()=>({error_message:""}));
			this.props.onSubmit({
				description:this.state.description,
				amount:parseFloat(this.state.amount,1) * 100,
				created_at:this.state.created_at.valueOf(),
				note:this.state.note
			});
		}
	}
	render(){
		return(
			<div>
				{this.state.error_message && <p>{this.state.error_message}</p>}
				<form onSubmit={this.on_submit}>
					<input
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.on_description_change}
					/>
					<input
						type="text"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.on_amount_change}
					/>
					<SingleDatePicker 
						date={this.state.created_at}
						onDateChange={this.on_date_change}
						focused={this.state.calendar_focused}
						onFocusChange={this.on_focus_change}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Add a not for youre expense(optional)"
						value={this.state.note}
						onChange={this.on_note_change}
					>
					</textarea>
					<button>Add Expense</button>
				</form>
			</div>
		);
	};
}