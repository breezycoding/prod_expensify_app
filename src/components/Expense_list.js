import React from "react";
import { connect } from "react-redux";
import Expense_list_item from "./Expense_list_item";
import select_expenses from "../selectors/expenses";
/*
	steps
	1. creating const for higher order component(HOC), in this case "Connected_expense_list"
		connect()()
			has two parameters
				1. the provide the information of what we want to connect/store information
				2. component you want to connect
				
	infos
	-map_state_to_props: will automatically rerun once the info in store changes
	-when you connect component to redux store its reactive, which means as the store changes your component will be rerendered with new values
		-all we have to do is define how we rerender things
*/
export const Expense_list = (props) => (
	<div>
		<h1>Expense List</h1>
		{
			props.expenses.length === 0 ? (
				<p>no expenses array</p>
			) : (
				props.expenses.map((expense)=>{
					return <Expense_list_item key={expense.id} {...expense}/>;
				})
			)
		}
	</div>
);

const map_state_to_props = (state) => {
	return{
		expenses:select_expenses(state.expenses, state.filters)
	};
};

export default connect(map_state_to_props)(Expense_list);
