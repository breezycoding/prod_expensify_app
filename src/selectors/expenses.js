import moment from "moment";

/*Get visible expenses*/
export default (expenses, {text, sort_by, start_date, end_date}) => {
	return expenses.filter((expense)=> {
		const created_at_moment = moment(expense.created_at);
		const start_date_match = start_date ? start_date.isSameOrBefore(created_at_moment, "day") : true;
		const end_date_match =  end_date ? end_date.isSameOrAfter(created_at_moment, "day") : true;
		const text_match = expense.description.toLowerCase().includes(text.toLowerCase());
		
		return start_date_match && end_date_match && text_match;
	}).sort((a, b) => {
		if(sort_by === "date"){
			return a.created_at < b.created_at ? 1 : -1;		
		}else if(sort_by === "amount"){
			return a.amount < b.amount ? 1 : -1;	 
			
		} 
	});
	
}