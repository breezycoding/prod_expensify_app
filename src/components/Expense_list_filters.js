import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates"; 
import 'react-dates/lib/css/_datepicker.css';
import { set_text_filter , sort_by_date, sort_by_amount, set_start_date, set_end_date } from "../actions/filters";

export class Expense_list_filters extends React.Component{
	state = {
		calendar_focused: null
	}
	onDatesChange = ({startDate, endDate}) => {
		this.props.set_start_date(startDate);
		this.props.set_end_date(endDate);
	};
	onFocusChange = (calendar_focused) => {
		this.setState(() => ({calendar_focused}) );
	};

	on_text_change = (event) => {
		this.props.set_text_filter(event.target.value);
	};
	
	on_sort_change = (event) =>{
		if(event.target.value === "date"){
			this.props.sort_by_date();
		}else if(event.target.value === "amount"){
			this.props.sort_by_amount(); 
		}
	};
	
	render(){
		return(
			<div>
				<input type="text" 
					value={this.props.filters.text} 
					onChange={this.on_text_change}
				/>
				<select
					value={this.props.filters.sort_by}
					onChange={this.on_sort_change}
				>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.start_date}
					startDateId={"startDateId"}
					endDate={this.props.filters.end_date}
					endDateId={"endDateId"}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendar_focused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		)
	}
}
const map_state_to_props =(state)=>{
	return{
		filters: state.filters	
	};
};
const mapDispatchToProps = (dispatch) => ({
	set_text_filter: (text) => dispatch(set_text_filter(text)),
	sort_by_date: () => dispatch(sort_by_date()),
	sort_by_amount: () => dispatch(sort_by_amount()),
	set_start_date: (startDate) => dispatch(set_start_date(startDate)),
	set_end_date: (endDate) => dispatch(set_end_date(endDate))
});


export default connect(map_state_to_props, mapDispatchToProps)(Expense_list_filters);