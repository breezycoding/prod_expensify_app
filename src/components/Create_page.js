import React from "react";
import { connect } from "react-redux";
import Expense_form from "./Expense_form";
import { add_expense } from "../actions/expenses";

export class Create_page extends React.Component{
	onSubmit = (expense) => {
		this.props.add_expense(expense);
		this.props.history.push("/");
	};
	
	render(){
		return(
			<div>
				<h1>Add Expense</h1>
				<Expense_form 
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}


const mapDispatchToProps = (dispatch) => ({
	add_expense: (expense) => dispatch(add_expense(expense))
});
export default connect(undefined, mapDispatchToProps)(Create_page);