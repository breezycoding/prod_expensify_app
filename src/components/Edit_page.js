import React from "react";
import { connect } from "react-redux";
import Expense_form from "./Expense_form";
import {edit_expense, remove_expense} from "../actions/expenses";

export class Edit_page extends React.Component{
	on_submit_edit = (expense) => {
		this.props.editExpense(this.props.expense.id, expense);
		this.props.history.push("/");
	}
	
	on_submit_remove = () => {
		this.props.removeExpense({ id:this.props.expense.id });
		this.props.history.push("/");
	}
	
	render(){
		return(
			<div>
				<Expense_form 
					expense={this.props.expense}
					onSubmit={this.on_submit_edit}
				/>
				<button onClick={this.on_submit_remove}>
					Remove
				</button>
			</div>
		)
	}
}

const map_to_state_props = (state, props) => {
	return {
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
	};
};

const mapDispatchToProps = (dispatch, props) => ({
	editExpense:(id, expense) => dispatch(edit_expense(id, expense)),
	removeExpense:(data) => dispatch(remove_expense(data))
});

export default connect(map_to_state_props, mapDispatchToProps)(Edit_page);