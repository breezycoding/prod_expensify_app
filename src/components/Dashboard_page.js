import React from "react";
import Expense_list from "./Expense_list";
import Expense_list_filters from "./Expense_list_filters";

const Dashboard_page = () => (
	<div>
		<Expense_list_filters />
		<Expense_list />
	</div>
);

export default Dashboard_page;