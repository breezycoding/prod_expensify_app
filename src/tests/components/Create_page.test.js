import React from "react";
import { shallow } from "enzyme";
import {Create_page} from "../../components/Create_page";
import expenses from "../fixtures/expenses";

let add_expense, history, wrapper;
beforeEach(() => {
	add_expense = jest.fn();
	history = { push: jest.fn() };
	wrapper = shallow(<Create_page add_expense={add_expense} history={history}/>);
});

test("should render Create_page correctly", () => {
	expect(wrapper).toMatchSnapshot();
});

test("should handle onSubmit", () => {
	wrapper.find("Expense_form").prop("onSubmit")(expenses[1]);
	expect(history.push).toHaveBeenLastCalledWith("/");
	expect(add_expense).toHaveBeenLastCalledWith(expenses[1]);
});

