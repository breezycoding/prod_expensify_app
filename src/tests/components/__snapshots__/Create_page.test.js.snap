// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Create_page correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <Expense_form
    onSubmit={[Function]}
  />
</div>
`;
